contracts/farm/Farm.sol

```diff
@@ -0,0 +1,197 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+import "./libraries/SafeToken.sol";
+import "./IFarm.sol";
+import "../core/interfaces/IERC20.sol";
+
+contract Farm is IFarm, OwnableUpgradeable, PausableUpgradeable {
+    using SafeToken for address;
+
+    struct StakeInfo {
+        uint256 amount;
+        uint256 rewardDebt;
+    }
+
+    struct PoolInfo {
+        uint256 allocPoint;
+        uint256 lastRewardBlock;
+        uint256 accRewardPerShare;
+    }
+
+    address public rewardToken;
+
+    uint256 public rewardPerBlock;
+    uint256 public totalAllocPoint;
+
+    mapping(address => PoolInfo) public poolInfos;
+    mapping(address => mapping(address => StakeInfo)) public stakes;
+    mapping(address => uint256) private lpTokenBalances;
+
+    event LogPoolAdded(address indexed lpToken, uint256 allocPoint);
+    event LogPoolAllocationUpdate(address indexed lpToken, uint256 allocPoint);
+    event LogPoolRewardUpdate(
+        address indexed lpToken,
+        uint256 accRewardPerShare,
+        uint256 lastRewardBlock
+    );
+    event LogDeposit(address indexed user, address indexed lpToken, uint256 amount);
+    event LogWithdraw(address indexed user, address indexed lpToken, uint256 amount);
+    event LogRewardPaid(address indexed user, address indexed lpToken, uint256 reward);
+
+    function initialize(
+        address _rewardToken,
+        uint256 _rewardPerBlock
+    ) external initializer {
+        OwnableUpgradeable.__Ownable_init();
+        PausableUpgradeable.__Pausable_init();
+
+        rewardToken = _rewardToken;
+        rewardPerBlock = _rewardPerBlock;
+        totalAllocPoint = 0;
+    }
+
+    function addPool(
+        address _lpToken,
+        uint256 _allocPoint
+    ) external override onlyOwner whenNotPaused {
+        poolInfos[_lpToken] = PoolInfo({
+            allocPoint: _allocPoint,
+            lastRewardBlock: block.number,
+            accRewardPerShare: 0
+        });
+
+        totalAllocPoint += _allocPoint;
+
+        emit LogPoolAdded(_lpToken, _allocPoint);
+    }
+
+    function updateAllocation(
+        address _lpToken,
+        uint256 _allocPoint
+    ) external override onlyOwner whenNotPaused {
+        PoolInfo storage pool = poolInfos[_lpToken];
+
+        totalAllocPoint = totalAllocPoint + _allocPoint - pool.allocPoint;
+        pool.allocPoint = _allocPoint;
+
+        emit LogPoolAllocationUpdate(_lpToken, _allocPoint);
+    }
+
+    function deposit(address _lpToken, uint256 _amount) external override whenNotPaused {
+        require(_lpToken != address(0), "Zero address");
```

Comment:
check for being contract is must, additionally if thats tech possible - check for interface